File: rxbinding/src/main/java/com/jakewharton/rxbinding2/widget/TextViewEditorActionEventObservable.java
package com.jakewharton.rxbinding2.widget;

import android.view.KeyEvent;
import android.widget.TextView;
import android.widget.TextView.OnEditorActionListener;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;
import io.reactivex.functions.Predicate;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class TextViewEditorActionEventObservable extends Observable<TextViewEditorActionEvent> {
  private final TextView view;
  private final Predicate<? super TextViewEditorActionEvent> handled;

  TextViewEditorActionEventObservable(TextView view,
      Predicate<? super TextViewEditorActionEvent> handled) {
    this.view = view;
    this.handled = handled;
  }

  @Override
  protected void subscribeActual(Observer<? super TextViewEditorActionEvent> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer, handled);
    observer.onSubscribe(listener);
    view.setOnEditorActionListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements OnEditorActionListener {
    private final TextView view;
    private final Observer<? super TextViewEditorActionEvent> observer;
    private final Predicate<? super TextViewEditorActionEvent> handled;

    Listener(TextView view, Observer<? super TextViewEditorActionEvent> observer,
        Predicate<? super TextViewEditorActionEvent> handled) {
      this.view = view;
      this.observer = observer;
      this.handled = handled;
    }

    @Override
    public boolean onEditorAction(TextView textView, int actionId, KeyEvent keyEvent) {
      TextViewEditorActionEvent event = TextViewEditorActionEvent.create(view, actionId, keyEvent);
      try {
        if (!isDisposed() && handled.test(event)) {
          observer.onNext(event);
          return true;
        }
      } catch (Exception e) {
        observer.onError(e);
        dispose();
      }
      return false;
    }

    @Override
    protected void onDispose() {
      view.setOnEditorActionListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding4/widget/TextViewEditorActionEventObservable.kt
@file:JvmName("RxTextView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding4.widget

import android.view.KeyEvent
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding4.internal.AlwaysTrue
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

import com.jakewharton.rxbinding4.internal.checkMainThread

/**
 * Create an observable of editor action events on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Warning:* The created observable uses [TextView.OnEditorActionListener] to
 * observe actions. Only one observable can be used for a view at a time.
 *
 * @param handled Predicate invoked each occurrence to determine the return value of the
 * underlying [TextView.OnEditorActionListener].
 */
@CheckResult
@JvmOverloads
fun TextView.editorActionEvents(
  handled: (TextViewEditorActionEvent) -> Boolean = AlwaysTrue
): Observable<TextViewEditorActionEvent> {
  return TextViewEditorActionEventObservable(this, handled)
}

data class TextViewEditorActionEvent(
  /** The view from which this event occurred.  */
  val view: TextView,
  val actionId: Int,
  val keyEvent: KeyEvent?
)

private class TextViewEditorActionEventObservable(
  private val view: TextView,
  private val handled: (TextViewEditorActionEvent) -> Boolean
) : Observable<TextViewEditorActionEvent>() {

  override fun subscribeActual(observer: Observer<in TextViewEditorActionEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer, handled)
    observer.onSubscribe(listener)
    view.setOnEditorActionListener(listener)
  }

  private class Listener(
    private val view: TextView,
    private val observer: Observer<in TextViewEditorActionEvent>,
    private val handled: (TextViewEditorActionEvent) -> Boolean
  ) : MainThreadDisposable(), OnEditorActionListener {

    override fun onEditorAction(textView: TextView, actionId: Int, keyEvent: KeyEvent?): Boolean {
      val event = TextViewEditorActionEvent(view, actionId, keyEvent)
      try {
        if (!isDisposed && handled(event)) {
          observer.onNext(event)
          return true
        }
      } catch (e: Exception) {
        observer.onError(e)
        dispose()
      }

      return false
    }

    override fun onDispose() {
      view.setOnEditorActionListener(null)
    }
  }
}
